### ISSUE TITLE
[zmq] pub/sub is not reliable at all

### ISSUE THREAD

**Issue by bitkevin** [bitcoin/bitcoin#12754]
The client only subscribe `hashblock` topic, if there's no block found over than about 30 minutes on the network, the pub/sub will timeout and the client can't detect anything unusual, the client will keep waiting forever.

I fix this by reconnecting, example code:

```cpp
#define BITCOIND_ZMQ_HASHBLOCK      "hashblock"

const time_t KReconnectInterval = 600;  // seconds

//
// check if need to reconnect ZMQ
//
if (subscriber == nullptr || lastRecvTime + KReconnectInterval < time(nullptr)) {
  // disconnect
  if (subscriber != nullptr) {
    delete subscriber;
    subscriber = nullptr;
  }

  // connect
  subscriber = new zmq::socket_t(zmqContext_, ZMQ_SUB);
  subscriber->connect(zmqBitcoindAddr_);

  // subscribe topic
  subscriber->setsockopt(ZMQ_SUBSCRIBE,
                         BITCOIND_ZMQ_HASHBLOCK, strlen(BITCOIND_ZMQ_HASHBLOCK));
}

//
// recv ZMQ messages
//

// ...

// set last recv time
lastRecvTime = time(nullptr);

// ...
```

I am not sure if add these options by `zmq_setsockopt()` could solve this issue.

```
// ZMQ_HEARTBEAT_xxxx available since zmq-4.2.0
ZMQ_HEARTBEAT_TTL
ZMQ_HEARTBEAT_TIMEOUT
ZMQ_HEARTBEAT_IVL
```

Maybe could add some heartbeat messages when there's no traffic for a while? I prefer this way.

Some links: 
* https://stackoverflow.com/questions/26112992/zeromq-pub-sub-on-unreliable-connection
* https://github.com/zeromq/jeromq/issues/364

**Comment by TheBlueMatt** [comment]
ZMQ_HEARTBEAT_* look promising, can you test with those in your test setup?

**Comment by luke-jr** [comment]
ZMQ is itself unreliable. I'm not sure this is a Core issue.

**Comment by ch4ot1c** [comment]
Looks like they're starting to sort it out in the ZMQ repo:

https://github.com/zeromq/libzmq/commit/0867c380326829dc7e48e12d46b977bffad207c4
https://github.com/zeromq/libzmq/commit/79d5ac3deeb944c9fd8a7a7a20a8b973d119fa5e
https://github.com/zeromq/libzmq/commit/cdf556610812c172a15c53da063ffd5684c5d995

We should probably await a fresh release / sufficient tests, but can begin on the implementation with 4.2.0 (using `zmq_setsockopt` and `ZMQ_HEARTBEAT_IVL`/`TIMEOUT`/`TTL`. I'll take a crack at it this week.

http://api.zeromq.org/4-2:zmq-setsockopt#toc17

**Comment by mruddy** [comment]
@ch4ot1c Are you able to reproduce the issue using the same tools from the bitcoin repo that I used in my test?

**Comment by ch4ot1c** [comment]
@mruddy Just tried it out, got the same results - everything seems to still be connected and appear immediately (block #7, after a `generate 1` after 45 minutes, port 28332):

<img width="791" alt="screen shot 2018-08-20 at 1 00 24 pm" src="https://user-images.githubusercontent.com/2287825/44357995-4460e900-a479-11e8-8efe-db8349c71f0f.png">

Maybe this is a timeout due to a middle box topology, like you mentioned?

**Comment by mruddy** [comment]
@ch4ot1c Thanks for confirming what you see!

I'm convinced that this behavior is caused by middle boxes. The OP could confirm with a little command line inspection using tools like `netstat`, `ss`, and `tcpdump`. He could verify that both sides think the connection is established and that some packets are sent, but not received, when this condition occurs.

With the middle box idea in mind, I've reproduced this behavior with a topology spread across multiple hosts with multiple NATs and firewalls in between.

The thing is that the ZMQ PUB sockets are not being created with TCP keep-alive enabled (i.e.- SO_KEEPALIVE, or more specifically to libzmq, setting `ZMQ_TCP_KEEPALIVE` to 1 in `CZMQAbstractPublishNotifier::Initialize`) This is easily verified with th the `-o` option to `netstat` or `ss`. I think enabling TCP keep-alive is what we want to do here because we want to keep the connection alive and we are not concerned if the other side is alive (what app level heartbeating is designed to do). Plus, using the TCP keep-alive options means we don't have to bump the min required library version (the heartbeat options are newer, see https://github.com/zeromq/libzmq/releases/tag/v4.2.0).

So, one way we can go is to set `ZMQ_TCP_KEEPALIVE` to 1 and then use the operating system provided defaults for keep alive interval and probes to fail before killing the connection. That works for me when the OS settings are set low enough to keep the middle boxes from forgetting the connection. The values must be set before starting the bitcoin process.

The other way we could go is to set `ZMQ_TCP_KEEPALIVE` to 1 and then also provide bitcoin command line args to customize the other ZMQ ZMQ_TCP_KEEPALIVE_* socket options (see http://api.zeromq.org/4-2:zmq-setsockopt).

For more on the OS keep alive settings see `/proc/sys/net/ipv4/tcp_keepalive_time`, `/proc/sys/net/ipv4/tcp_keepalive_intvl`, and `/proc/sys/net/ipv4/tcp_keepalive_probes` in `man 7 tcp`.

**Comment by bitkevin** [comment]
After read zmq doc, still have no idea how to use `ZMQ_HEARTBEAT_*`, copy some code snippets, try use `zmq_socket_monitor()`, heartbeat still not work, no timeout event or similar event.

But reconnect after no message in a while is very stable.

**Comment by mruddy** [comment]
`ZMQ_HEARTBEAT_*` is not what you want. i'll make some `ZMQ_TCP_KEEPALIVE` changes available after some of my other zmq stuff gets merged to avoid having to rebase more than necessary.

**Comment by adamjonas** [comment]
Closed by #14687.

**Comment by MarcoFalke** [comment]
Is this still an issue with a recent version of Bitcoin Core? If yes, what are the steps to reproduce?
